### PR TITLE
Fixing Distributed tracing call flow

### LINKED ISSUE #1373: Incorrect distributed tracing call flow
# In what area(s)?

/area runtime

## What version of Dapr?

0.6.0

### Aggregation per component and dependency services are not working on AppInsights

The below application map draws the service invocation traces between pythonapp and nodeapp

image

1. nodeapp calls /daprinternal/CallLocal (`S3`): This is wrong because daprinternal/CallLocal represents the internal gRPC channel in pythonapp and nodeapp. So this should not be represented as the actual service. Technically, the latency(5.5ms) between pythonapp and nodeapp should be same as the latency(3.6ms) between nodeapp and CallLocal, but it is not.
2. nodeapp calls /127.0.0.1/neworder (`S2`): is small circle “dependency service” ? this circle may be represented as an user app
3. nodeapp calls /v1.0/state/statestore (`S4`): statestore must be a dependency service, but the name of this small circle makes user confused because big circle (nodeapp) can be treated as user app. Then this is conflicted with the assumption of `S4` circle
4. nodeapp calls /v1.0/healthz (`S1`): this is called from kubernetes. we should remove this healthz call

In this diagram, small circle can be treated as dapr side car or user app. This makes user confused to understand the application map. It also shows the inconsistency of the latency. 

When we click the big circle(`B2`), I am expecting to see the list of request coming from the dependency services such as CallLocal from pythonapp and user app, but we see nothing. So aggregation per component or dependency does not work. 

### Incorrect server/client spans on zipkin.

image

* Arrow Marked call should be called from PYTHONAPP, not NODEAPP
* Server span of the last two NODEAPP spans is unclear. We do not need to generate serverspan on app channel.

## Proposal

Defining the big circle and small circle which are shown in application insights map.

* Big circle: Dapr side car, Dapr system components
* Small circle: external dependency (statestore, bindings, pubsub) or user app

> Note: this will show Dapr side car and user app separately. 

image

ServerSpan and ClientSpan per server and dependency will be aggregated like the below examples. The aggregated api calls, including internal gRPC method call and /v1.0/state/statestore will be shown when user clicks the circle. 

image

> Note: from the investigation, if we resolve application insight map issues, we can resolve zipkin tracing issues as well.

### PR DESCRIPTION
**Pull request by shalabhms**
# Description

Changes to fix distributed trace flows . reference #1373 

## Issue reference

Closes #1373 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[https://github.com/dapr/docs/issues/550]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/diagnostics/grpc_tracing_test.go [hunk 1]
```diff
@@ -7,20 +7,40 @@ package diagnostics
...
+	}
+	wantOk := true
+	ctx = AppendToOutgoingGRPCContext(ctx, wantSc)
+
+	gotSc, gotOk := FromOutgoingGRPCContext(ctx)
 
-	// TODO: check if the dapr-headerKey is populated to span annotation
+	if !reflect.DeepEqual(gotSc, wantSc) {
+		t.Errorf("WithGRPCSpanContext gotSc = %v, want %v", gotSc, wantSc)
+	}
```
Comment (youngbupark) [inline pkg/diagnostics/grpc_tracing_test.go:42]: We can use `assert.Equalf`
```
assert.NotEqualf(t, gotSc, wantSc, "WithGRPCSpanContext gotSc = %v, want %v", gotSc, wantSc)
```

File: pkg/diagnostics/grpc_tracing_test.go [hunk 2]
```diff
@@ -7,20 +7,40 @@ package diagnostics
...
+	if gotOk != wantOk {
+		t.Errorf("WithGRPCSpanContext gotOk = %v, want %v", gotOk, wantSc)
+	}
```
Comment (youngbupark) [inline pkg/diagnostics/grpc_tracing_test.go:45]: this is not required.

File: pkg/diagnostics/http_tracing.go [hunk 3]
```diff
@@ -7,70 +7,140 @@ package diagnostics
```
Comment (youngbupark) [inline pkg/diagnostics/http_tracing.go:7]: We need to describe where we got the code..

File: pkg/diagnostics/http_tracing_test.go [hunk 4]
```diff
@@ -6,27 +6,117 @@
...
+			req.Header.Add("traceparent", tt.header)
 
-	TraceSpanFromFastHTTPContext(ctx, spec)
+			gotSc, gotOk := SpanContextFromRequest(req)
+			if !reflect.DeepEqual(gotSc, tt.wantSc) {
+				t.Errorf("SpanContextFromRequest() gotSc = %v, want %v", gotSc, tt.wantSc)
+			}
+			if gotOk != tt.wantOk {
+				t.Errorf("SpanContextFromRequest gotOk = %v, want %v", gotOk, tt.wantOk)
+			}
```
Comment (youngbupark) [inline pkg/diagnostics/http_tracing_test.go:90]: we can use assert.NotEqualf for the replacement of L84-L90

File: pkg/diagnostics/http_tracing_test.go [hunk 5]
```diff
@@ -6,27 +6,117 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.wantHeader, func(t *testing.T) {
+			req := &fasthttp.Request{}
+			SpanContextToRequest(tt.sc, req)
+
+			h := string(req.Header.Peek("traceparent"))
+			if got, want := h, tt.wantHeader; got != want {
+				t.Errorf("SpanContextToRequest() header = %v, want %v", got, want)
```
Comment (youngbupark) [inline pkg/diagnostics/http_tracing_test.go:116]: please use assert.xxx

File: pkg/diagnostics/tracing.go [hunk 6]
```diff
@@ -39,17 +49,43 @@ func SerializeSpanContext(ctx trace.SpanContext) string {
 
 // DeserializeSpanContext deserializes a span context from a string
 func DeserializeSpanContext(ctx string) trace.SpanContext {
+	var ret trace.SpanContext
```
Comment (youngbupark) [inline pkg/diagnostics/tracing.go:52]: Do we still need this ?

**Comment by shalabhms** [comment]
image

File: pkg/diagnostics/tracing.go (hunk 6 above)
Comment (shalabhms) [reply pkg/diagnostics/tracing.go]: no. and I removed related codes in similar lines.

**Pull request merged**